python/ray/train/callbacks/results_preprocessors/average.py

```diff
@@ -0,0 +1,104 @@
+import logging
+from typing import Dict, List, Tuple
+
+import numpy as np
+
+from ray.train.callbacks.results_preprocessors.preprocessor import ResultsPreprocessor
+
+
+logger = logging.getLogger(__name__)
+
+
+class AverageResultsPreprocessor(ResultsPreprocessor):
```

Comment:
These two comments totally make sense to me. I think it depends what we allow users to do with this result preprocessor. Some impacts of appending a single `_aggregated_metrics` to the `results` list:
1. The `results` length is no longer the number of workers. 
2. Users need to know which position it is appended (usually the end) to use it for callbacks (e.g. TensorBoard).
3. Some default metrics are not very meaningful in this new entry, such as timestamp and node ID.

Also, should we prevent users from using this average preprocessor as a callback preprocessor?

Reply:
Yeah, I think it would be good to make it more generic and not tie it down to one type of operation. I liked the idea in the initial draft of the PR, with the user specifying their own functions.

As for the issues with appending the results, here is what I would suggest - subclass `list` (`ResultsList`?)  and have `aggregate` be an additional property that can also be accessed by `obj["aggregate"]`. That way, it is still possible to do all of the list operations and the length will match the number of workers. The downside is that it may be a little confusing for the user, but eg. Skorch implements a similar approach. I wouldn't worry about some metrics not being meaningful.

Reply:
> Yeah, I think it would be good to make it more generic and not tie it down to one type of operation. I liked the idea in the initial draft of the PR, with the user specifying their own functions.

So we don't need to have a specific averaging results preprocessor, just an aggregation proprocessor. Users can define their own aggregators using reported metrics from workers. Of course, we can provide users with some averaging methods.

> As for the issues with appending the results, here is what I would suggest - subclass `list` (`ResultsList`?) and have `aggregate` be an additional property that can also be accessed by `obj["aggregate"]`. That way, it is still possible to do all of the list operations and the length will match the number of workers. The downside is that it may be a little confusing for the user, but eg. Skorch implements a similar approach. I wouldn't worry about some metrics not being meaningful.

This sounds a good idea to me. Should we add some doc and update existing callbacks to tell users how they can use the aggregated metrics?

Reply:
Ah shoot... I do think this is quite a clever solution but my major concern is about how the user would be able to interact with this entity within an individual Callback and more generally how we would extend the existing Callbacks to support this. Currently they just iterate through the list - one naive idea is to have some logic to fetch `obj["aggregate"]` and append it _after_ the `_preprocess_results` step, but that doesn't seem like an ideal user experience either.

Reply:
> Ah shoot... I do think this is quite a clever solution but my major concern is about how the user would be able to interact with this entity within an individual Callback and more generally how we would extend the existing Callbacks to support this. Currently they just iterate through the list - one naive idea is to have some logic to fetch `obj["aggregate"]` and append it _after_ the `_preprocess_results` step, but that doesn't seem like an ideal user experience either.

I have similar concerns. Does the current implementation seem to be a reasonable compromise? (i.e. appending aggregated metrics to the result Dict of each worker. So they are visible to all, and there is no need for adjustments in callbacks. The drawback is that it may be redundant.)

I will add the support for generic aggregation methods and default averaging first. We can discuss more on where to store these aggregated results.

Reply:
Yeah doing the generic aggregation first makes sense, I think we're all in agreement there!

As for where to place the aggregates, I'm leaning towards expanding on the `obj["aggregate"]` idea and changing the original results aggregated to be a more concrete dataclass or even a dict.  By default you might get something like:

`{"raw": [worker0_results, worker1_results, ...]}`

And then preprocessors can instead operate on the entire `dict` and append new keys (e.g. `"aggregate"`).

Haven't fully fleshed out how to completely merge these at the end at the individual `Callback` level, but I think this type of composability would be the right start. cc @amogkam for more input